File: agent_memory_mcp/engine/migration.py
"""One-time import of legacy markdown memory files into SQLite."""

from __future__ import annotations

import json
import re
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .db import MemoryEngine

ENTRY_RE = re.compile(r"^### (.+)$", re.MULTILINE)


def import_legacy_markdown(engine: "MemoryEngine", memory_dir: Path, repos_map: dict) -> int:
    """Import */memory/*.md and state.json failure_analytics. Returns count imported."""
    if not memory_dir.exists():
        return 0

    imported = 0
    for repo_dir in memory_dir.iterdir():
        if not repo_dir.is_dir():
            continue
        repo_id = repo_dir.name
        mem_path = repo_dir / "memory"
        if not mem_path.exists():
            continue

        path = repos_map.get(repo_id, str(repo_dir))
        engine.upsert_repo(repo_id, path)

        for kind_file, kind in [
            ("failures.md", "failure"),
            ("decisions.md", "decision"),
            ("attempts.md", "attempt"),
        ]:
            fpath = mem_path / kind_file
            if not fpath.exists():
                continue
            text = fpath.read_text(encoding="utf-8")
            blocks = _split_markdown_entries(text)
            for ts, body in blocks:
                if not body.strip():
                    continue
                _, inserted = engine.insert_memory(
                    repo_id,
                    kind,
                    body.strip(),
                    source="import",
                    metadata={"legacy_timestamp": ts},
                    skip_failure_dedup=True,
                )
                if inserted:
                    imported += 1

        state_path = mem_path / "state.json"
        if state_path.exists():
            try:
                state = json.loads(state_path.read_text())
                for sig, data in state.get("failure_analytics", {}).items():
                    engine._conn.execute(
                        """
                        INSERT OR IGNORE INTO failure_signatures
                        (repo_id, signature, count, first_seen, last_seen, resolved, memory_id)
                        VALUES (?, ?, ?, ?, ?, ?, NULL)
                        """,
                        (
                            repo_id,
                            sig,
                            data.get("count", 1),
                            data.get("first_seen", ""),
                            data.get("last_seen", ""),
                            1 if data.get("resolved") else 0,
                        ),
                    )
                engine._conn.commit()
            except (json.JSONDecodeError, OSError):
                pass

    marker = memory_dir.parent / ".migrated_to_sqlite"
    if imported > 0 or memory_dir.exists():
        marker.write_text("ok")
    return imported


def _split_markdown_entries(text: str) -> list[tuple[str, str]]:
    parts: list[tuple[str, str]] = []
    matches = list(ENTRY_RE.finditer(text))
    if not matches:
        if text.strip():
            parts.append(("", text))
        return parts
    for i, m in enumerate(matches):
        ts = m.group(1)
        start = m.end()
        end = matches[i + 1].start() if i + 1 < len(matches) else len(text)
        parts.append((ts, text[start:end]))
    return parts


def needs_migration(root: Path) -> bool:
    marker = root / ".migrated_to_sqlite"
    memory_dir = root / "agent-memory"
    return memory_dir.exists() and not marker.exists()

File: internal/store/sqlite/migration.go
package sqlite

import (
	"context"
	"encoding/json"
	"io/ioutil"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

var entryRe = regexp.MustCompile(`(?m)^### (.+)$`)

func RunLegacyMigration(ctx context.Context, s *Store, root string) error {
	markerPath := filepath.Join(root, ".migrated_to_sqlite")
	if _, err := os.Stat(markerPath); err == nil {
		return nil // Already migrated
	}

	memoryDir := filepath.Join(root, "agent-memory")
	if _, err := os.Stat(memoryDir); os.IsNotExist(err) {
		return nil // Nothing to migrate
	}

	reposMapPath := filepath.Join(root, "repos.json")
	reposMap := make(map[string]string)
	if b, err := ioutil.ReadFile(reposMapPath); err == nil {
		json.Unmarshal(b, &reposMap)
	}

	entries, err := ioutil.ReadDir(memoryDir)
	if err != nil {
		return err
	}

	imported := 0
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		repoID := entry.Name()
		memPath := filepath.Join(memoryDir, repoID, "memory")
		if _, err := os.Stat(memPath); os.IsNotExist(err) {
			continue
		}

		path, ok := reposMap[repoID]
		if !ok {
			path = filepath.Join(memoryDir, repoID)
		}
		s.UpsertRepo(ctx, repoID, path)

		kinds := map[string]string{
			"failures.md":  "failure",
			"decisions.md": "decision",
			"attempts.md":  "attempt",
		}

		for file, kind := range kinds {
			fpath := filepath.Join(memPath, file)
			b, err := ioutil.ReadFile(fpath)
			if err != nil {
				continue
			}

			blocks := splitMarkdownEntries(string(b))
			for _, block := range blocks {
				if strings.TrimSpace(block.Body) == "" {
					continue
				}
				meta := map[string]any{"legacy_timestamp": block.Timestamp}
				_, inserted, _ := s.InsertMemory(ctx, repoID, kind, strings.TrimSpace(block.Body), "import", meta, true)
				if inserted {
					imported++
				}
			}
		}

		statePath := filepath.Join(memPath, "state.json")
		if b, err := ioutil.ReadFile(statePath); err == nil {
			var state map[string]any
			if json.Unmarshal(b, &state) == nil {
				if fa, ok := state["failure_analytics"].(map[string]any); ok {
					for sig, dataAny := range fa {
						if data, ok := dataAny.(map[string]any); ok {
							count := 1
							if c, ok := data["count"].(float64); ok { count = int(c) }
							fs := ""
							if f, ok := data["first_seen"].(string); ok { fs = f }
							ls := ""
							if l, ok := data["last_seen"].(string); ok { ls = l }
							res := 0
							if r, ok := data["resolved"].(bool); ok && r { res = 1 }

							s.memoryDB.ExecContext(ctx, `
								INSERT OR IGNORE INTO failure_signatures
								(repo_id, signature, count, first_seen, last_seen, resolved, memory_id)
								VALUES (?, ?, ?, ?, ?, ?, NULL)
							`, repoID, sig, count, fs, ls, res)
						}
					}
				}
			}
		}
	}

	if imported > 0 || len(entries) > 0 {
		ioutil.WriteFile(markerPath, []byte("ok"), 0644)
	}

	return nil
}

type mdBlock struct {
	Timestamp string
	Body      string
}

func splitMarkdownEntries(text string) []mdBlock {
	var parts []mdBlock
	matches := entryRe.FindAllStringSubmatchIndex(text, -1)
	if len(matches) == 0 {
		if strings.TrimSpace(text) != "" {
			parts = append(parts, mdBlock{Timestamp: "", Body: text})
		}
		return parts
	}

	for i, m := range matches {
		ts := text[m[2]:m[3]]
		start := m[1]
		end := len(text)
		if i+1 < len(matches) {
			end = matches[i+1][0]
		}
		parts = append(parts, mdBlock{
			Timestamp: ts,
			Body:      text[start:end],
		})
	}
	return parts
}
